compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt

```diff
@@ -453,6 +461,66 @@ class ResolverImpl(
         return KSTypeArgumentLiteImpl.getCached(typeRef, variance)
     }
 
+    override fun asMemberOf(
+        property: KSPropertyDeclaration,
+        containing: KSType
+    ): KSType {
+        val propertyDeclaredIn = property.closestClassDeclaration()
+            ?: throw IllegalArgumentException("Cannot call asMemberOf with a property that is " +
+                "not declared in a class or an interface")
+        val declaration = resolvePropertyDeclaration(property)
+        if (declaration != null && containing is KSTypeImpl && !containing.isError) {
+            if (!containing.kotlinType.isSubtypeOf(propertyDeclaredIn)) {
+                throw IllegalArgumentException(
+                    "$containing is not a sub type of the class/interface that contains `$property` ($propertyDeclaredIn)"
+                )
+            }
+            val typeSubstitutor = containing.kotlinType.createTypeSubstitutor()
+            val substituted = declaration.substitute(typeSubstitutor) as? ValueDescriptor
+            substituted?.let {
+                return KSTypeImpl.getCached(substituted.type)
+            }
+        }
+        // if substitution fails, fallback to the type from the property
+        return KSErrorType
+    }
+
+    override fun asMemberOf(
+        function: KSFunctionDeclaration,
+        containing: KSType
+    ): KSFunction {
+        val functionDeclaredIn = function.closestClassDeclaration()
+            ?: throw IllegalArgumentException("Cannot call asMemberOf with a function that is " +
+                "not declared in a class or an interface")
+        val declaration = resolveFunctionDeclaration(function)
+        if (declaration != null && containing is KSTypeImpl && !containing.isError) {
+            if (!containing.kotlinType.isSubtypeOf(functionDeclaredIn)) {
+                throw IllegalArgumentException(
+                    "$containing is not a sub type of the class/interface that contains " +
+                        "`$function` ($functionDeclaredIn)"
+                )
+            }
+            val typeSubstitutor = containing.kotlinType.createTypeSubstitutor()
+            val substituted = declaration.substitute(typeSubstitutor)
+            return KSFunctionImpl(substituted)
+        }
```

Comment:
Just curious, what's the `KotlinType` returned for functions with their own type parameters? IIUC they cannot be expressed by `KotlinType`. E.g., `fun <T> g() = 0`.

On the other hand, if they are available in the form of `KotlinType`, then we probably don't need `KSFunction` and can just use `KSType`, which is underpinned by `KotlinType`.

Reply:
i'm not sure how to get that. `declaration` does not have a KotlinType in it. Is there a helper method to get that?

Reply:
Sorry it's my bad. Please ignore my comments. I was thinking `substituted` a `KotlinType`.